### PR TITLE
Feat/region trait

### PR DESCRIPTION
**Pull request by aborgna-q**
Adds a `Base` parameter to the views, so we can make a view from any `HugrView`. Most of the noise comes from adding the generic constraints everywhere.

The more useful part of these changes is the new trait
```rust
pub trait Region<'a, Base>:
    HugrView
    + ::petgraph::visit::GraphBase<NodeId = Node>
    + ::petgraph::visit::GraphProp
    + ::petgraph::visit::NodeCount
    + ::petgraph::visit::NodeIndexable
    + ::petgraph::visit::EdgeCount
    + ::petgraph::visit::Visitable
    + ::petgraph::visit::GetAdjacencyMatrix
    + ::petgraph::visit::Visitable
where
    for<'g> &'g Self:
        ::petgraph::visit::IntoNeighborsDirected + ::petgraph::visit::IntoNodeIdentifiers,
```

that allows us to generalize over views where we can run petgraph's algorithms.

### CODE REVIEW

File: src/hugr/region/petgraph.rs [hunk 1]
```diff
@@ -12,22 +13,34 @@ use portgraph::NodeIndex;
 
 macro_rules! impl_region_petgraph_traits {
     ($hugr:ident) => {
-        impl<'a> petgraph::visit::GraphBase for $hugr<'a> {
+        impl<'a, Base> petgraph::visit::GraphBase for $hugr<'a, Base>
```
Comment (acl-cqc) [inline src/hugr/region/petgraph.rs:16]: nit: I wonder about something like `use petgraph::visit as pv` for all of these (and maybe get all on one line with the constraints)

**Review by acl-cqc — APPROVED** [review]
This is great, thanks. Mutable regions to follow I presume? :-) :-)

File: src/hugr/view.rs [hunk 2]
```diff
@@ -251,7 +251,10 @@ pub(crate) mod sealed {
         type Portgraph: LinkView;
 
         /// Returns a reference to the underlying portgraph.
-        fn as_portgraph(&self) -> &Self::Portgraph;
+        fn portgraph(&self) -> &Self::Portgraph;
+
+        /// Returns a the Hugr at the base of a chain of views.
+        fn base_hugr(&self) -> &Hugr;
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply src/hugr/view.rs:257]: There's always a hugr at the base of the stack, so this is equivalent to returning each component on its own.

File: src/hugr/region.rs [hunk 3]
```diff
@@ -248,20 +236,20 @@ impl<'g> HugrView for RegionView<'g> {
...
+        self.hugr
+            .get_parent(node)
+            .filter(|&parent| self.graph.contains_node(parent.index))
             .map(Into::into)
     }
 
     #[inline]
     fn get_optype(&self, node: Node) -> &OpType {
-        self.op_types.get(node.index)
+        self.hugr.get_optype(node)
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply src/hugr/region.rs:249]: The behaviour when querying invalid nodes is not defined, so this is compliant with the trait.

We could filter and return a reference to a default operation instead, but that may be even more confusing (and you have to store that operation somewhere...).

File: src/hugr/region.rs [hunk 4]
```diff
@@ -326,6 +314,93 @@ impl<'g> HugrView for RegionView<'g> {
     }
 }
 
+/// A common trait for views of a hugr region.
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply src/hugr/region.rs:319]: I just did a small change to replace the `Region` type parameter with a `Base` associated type, as it ends up being cleaner.

I'd really like to replace the petgraph trait impl macro with blanket implementations using this, but unfortunately we cannot do that for foreign traits :/

**Pull request merged**